op-supervisor/supervisor/backend/cross/cycle_test.go

```diff
@@ -0,0 +1,511 @@
+	}
+	runHazardCycleChecksTestCaseGroup(t, "Failure", tests)
+}
+
+func TestHazardCycleChecksNoCycle(t *testing.T) {
+	tests := []hazardCycleChecksTestCase{
+		{
+			name:        "no logs",
+			chainBlocks: emptyChainBlocks,
+			expectErr:   nil,
+			msg:         "expected no cycle found for block with no logs",
+		},
+		{
+			name: "one basic log",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 1,
+					messages: map[uint32]*types.ExecutingMessage{},
+				},
+			},
+			msg: "expected no cycle found for single basic log",
+		},
+		{
+			name: "one exec log",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 0),
+					},
+				},
+			},
+			msg: "expected no cycle found for single exec log",
+		},
+		{
+			name: "two basic logs",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{},
+				},
+			},
+			msg: "expected no cycle found for two basic logs",
+		},
+		{
+			name: "two exec logs to same target",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 3,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 0),
+						2: execMsg("1", 0),
+					},
+				},
+			},
+			msg: "expected no cycle found for two exec logs pointing at the same log",
+		},
+		{
+			name: "two exec logs to different targets",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 3,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 0),
+						2: execMsg("1", 1),
+					},
+				},
+			},
+			msg: "expected no cycle found for two exec logs pointing at the different logs",
+		},
+		{
+			name: "one basic log one exec log",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 0),
+					},
+				},
+			},
+			msg: "expected no cycle found for one basic and one exec log",
+		},
+		{
+			name: "first log is exec",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 1,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("2", 0),
+					},
+				},
+				"2": {
+					logCount: 1,
+					messages: nil,
+				},
+			},
+			msg: "expected no cycle found first log is exec",
+		},
+		{
+			name: "cycle through different timestamp",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("2", 0),
+						1: execMsgWithTimestamp("2", 1, 99),
+					},
+				},
+				"2": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("1", 1),
+					},
+				},
+			},
+			msg: "expected no cycle detection error for cycle through messages with different timestamps",
+		},
+	}
+	runHazardCycleChecksTestCaseGroup(t, "NoCycle", tests)
+}
+
+func TestHazardCycleChecksCycle(t *testing.T) {
```

Comment:
A test case with a 3rd chain may be interesting, to show a cycle is detected, even if one of the chains isn't involved in the cycle.